Source/Particles/PhysicalParticleContainer.cpp

```diff
@@ -358,27 +358,49 @@ PhysicalParticleContainer::AddPlasmaFromFile(ParticleReal q_tot,
             weight = w * w_unit;
         }
 
+        // TODO: Add option for different levels of injection
+        const Geometry& geom = Geom(0);
+        const auto problo = geom.ProbLoArray();
+        const auto probhi = geom.ProbHiArray();
+
         for (auto i = decltype(npart){0}; i<npart; ++i){
             ParticleReal const x = ptr_x.get()[i]*position_unit_x;
-            ParticleReal const z = ptr_z.get()[i]*position_unit_z+z_shift;
+            if (x>problo[0] && x<probhi[0]){
+                ParticleReal const z = ptr_z.get()[i]*position_unit_z+z_shift;
 #   ifndef WARPX_DIM_3D
-            ParticleReal const y = 0.0_prt;
+                if (z>problo[1] && z<probhi[1]){
```

Comment:
I am not sure the if cascased here are logically stable. A particle should only be initialized if it satisfies all contrains (x,y,z), so maybe they should be `&&` concatenated, stored in a temporary boolean variable such as `is_valid` and then we initialize it?

Reply:
The problem is in 2D runs, this won't work unless you set the y to 0.

Reply:
You can check the beam definition above this function as an example.

Reply:
I am not sure I understood your point here very clearly.
As the code is, the particle is only initialized if it meets all the if constrains.
With the cascade option, if the particle fails to be inside in the x dimension, we don't loose time to check all others.

Reply:
I think the nested logic makes it hard to read this part. I suggest to create and intermediate variable instead of nesting multiple ifs inside a for loop. Let me add a code suggestion.

Reply:
Oh ok. Don't worry! I can try to do it.

Reply:
I opened the file locally now and see what you did: you try to avoid loading data when you already know the particle will not be initialized anyway. I feel this is probably an optimization that is not super critical, but if you like to add it maybe use blocks like these:
```C++
            bool const in_box_x = x>problo[0] && x<probhi[0];
            if( not in_x_bounds ) continue;
```
Instead of nesting the `ifs`. Otherwise I feel this gets unnecessarily hard to read, especially with the interleaved preprocessor commands.

Reply:
That sounds good! I will do that and test it, thanks @ax3l !